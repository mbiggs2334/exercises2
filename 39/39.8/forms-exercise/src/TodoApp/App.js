import React from 'react';

const App = () => {
    return (
        'TodoApp'
    )
};

export default App