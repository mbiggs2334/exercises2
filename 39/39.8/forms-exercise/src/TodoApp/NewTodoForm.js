import React, { useState } from 'react';

const NewTodoForm = ({removeTodo, addTodo}) => {
    const [input, setInput] = useState('');

    const handleChange = (e) => {
        const {value} = e.target;
        setInput(value)
    };

    const handleSubmit = (e) => {
        e.preventDefault();
        addTodo(input)
        setInput('');
    }

    return (
        <form onSubmit={handleSubmit}>
            <label htmlFor="todo">Todo: </label>
            <input
                type="text"
                id="todo"
                required
                placeholder="Enter a todo..."
                value={input}
                name="todo"
                onChange={handleChange}
            />
            <button>Add Todo</button>
        </form>
    )
};

export default NewTodoForm;