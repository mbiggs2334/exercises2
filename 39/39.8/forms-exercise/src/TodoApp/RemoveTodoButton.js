import React from 'react';

const RemoveTodoButton = ({removeTodo}) => {
    return (
        <button onClick={removeTodo}>X</button>
    )
};

export default RemoveTodoButton;