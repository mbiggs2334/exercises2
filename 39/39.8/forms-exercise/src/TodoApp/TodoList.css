.TodoList-list {
    text-align: left;
}