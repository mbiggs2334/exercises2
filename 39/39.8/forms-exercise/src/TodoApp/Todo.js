import React from 'react';
import RemoveTodoButton from './RemoveTodoButton';

const Todo = ({todo, removeTodo}) => {

    const handleRemove = (e) => {
        removeTodo(todo);
    }

    return (
        <li>{todo} <RemoveTodoButton removeTodo={handleRemove} /></li>
    )
};

export default Todo;