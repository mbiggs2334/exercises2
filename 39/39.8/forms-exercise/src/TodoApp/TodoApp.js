import React from 'react';
import TodoList from './TodoList';
import './TodoApp.css'

const TodoApp = () => {
    return (
        <div className="TodoApp">
            <h1>Make a nice TodoList</h1>
            <h2>Don't you just hate it when you forget things?</h2>
            <h3>Now you don't have to!</h3>
            <TodoList />
        </div>
    )
};

export default TodoApp