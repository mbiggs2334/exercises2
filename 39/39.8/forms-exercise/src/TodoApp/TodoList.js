import React, { useState } from 'react';
import { v4 as uuid } from 'uuid';
import Todo from './Todo';
import NewTodoForm from './NewTodoForm';
import './TodoList.css';

const TodoList = () => {

    const [todoList, setTodoList] = useState([]);

    const removeTodo = (todo) => {
        const idx = todoList.indexOf(todo);
        todoList.splice(idx, 1);
        setTodoList([...todoList]);
    };

    const addTodo = (newTodo) => {
        setTodoList([...todoList, newTodo]);
    };

    return (
        <>
            <NewTodoForm addTodo={addTodo} />
            <div className="TodoList">
                <ul className='TodoList-list'>
                    {todoList.map(val => 
                        <Todo key={uuid()} todo={val} removeTodo={removeTodo} />
                    )}
                </ul>
            </div>
        </>
        
    )
};

export default TodoList;