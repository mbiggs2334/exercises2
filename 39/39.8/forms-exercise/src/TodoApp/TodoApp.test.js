import React from "react";
import { render, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import TodoApp from "./TodoApp";

it('renders without crashing', () => {
    render(<TodoApp />);
});

it('matches the snapshot', () => {
    const { asFragment } = render( <TodoApp />);
    expect(asFragment()).toMatchSnapshot();
});


it('should contain headers', () => {
    const { queryByText } = render(<TodoApp />);

    expect(queryByText("Make a nice TodoList")).toBeInTheDocument();
});

it('should add/remove a todo', () => {
    const { queryByText, queryByLabelText } = render(<TodoApp />);

    expect(queryByText('test app')).not.toBeInTheDocument();

    const input = queryByLabelText("Todo:");
    const button = queryByText('Add Todo');
    fireEvent.change(input, {target: {value: 'test app'}});
    fireEvent.click(button);

    expect(queryByText('test app')).toBeInTheDocument();

    const removeButton = queryByText('X');
    fireEvent.click(removeButton);
    expect(queryByText('test app')).not.toBeInTheDocument();
});