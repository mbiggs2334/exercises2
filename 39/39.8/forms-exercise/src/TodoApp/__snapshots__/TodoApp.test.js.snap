// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoApp"
  >
    <h1>
      Make a nice TodoList
    </h1>
    <h2>
      Don't you just hate it when you forget things?
    </h2>
    <h3>
      Now you don't have to!
    </h3>
    <form>
      <label
        for="todo"
      >
        Todo: 
      </label>
      <input
        id="todo"
        name="todo"
        placeholder="Enter a todo..."
        required=""
        type="text"
        value=""
      />
      <button>
        Add Todo
      </button>
    </form>
    <div
      class="TodoList"
    >
      <ul
        class="TodoList-list"
      />
    </div>
  </div>
</DocumentFragment>
`;
